File: slides/code_examples/cpp_examples/src/data_race.cpp
#include <iostream>
#include <thread>

int counter = 0;

void increment() {
  for (int i = 0; i < 10000; i++) {
    counter++; // Data race here
  }
}

int main() {
  std::thread t1(increment);
  std::thread t2(increment);

  t1.join();
  t2.join();

  std::cout << "Counter: " << counter << std::endl;
  return 0;
}

File: slides/code_examples/rust_examples/src/bin/data_race.rs
use std::thread;

static mut COUNTER: i32 = 0;

fn main() {
    let increment = move || {
        for _ in 0..10_000 {
            unsafe {
                // doesn't compile without
                COUNTER += 1;
            }
        }
    };

    let t1 = thread::spawn(increment);
    let t2 = thread::spawn(increment);

    t1.join().unwrap();
    t2.join().unwrap();

    unsafe {
        // doesn't compile without
        println!("Counter: {}", COUNTER);
    }
}
